OmsAgent/watcherutil.py

```diff
@@ -162,3 +251,261 @@ def watch(self):
+            self._consecutive_restarts_due_to_error = 0       
+
+    def monitor_resource(self, self_mon_info):
+        """
+            Monitor resource utilization of omsagent.
+            Check for memory and CPU periodically. If they cross the threshold for consecutive 3 iterations
+            we will restart the agent. 
+        """
+
+        resource_usage = self.get_oms_agent_resource_usage()
+        message = "Memory : {0}, CPU : {1}".format(resource_usage[0], resource_usage[1])
+        event = self.create_telemetry_event("agenttelemetry","True",message,"300000")        
+        self.write_waagent_event(event)
+
+        self_mon_info._memory_used_in_percent = resource_usage[0]
+                
+        if (self_mon_info._memory_used_in_percent > 0):
+            if (self_mon_info._memory_used_in_percent > MemoryThresholdToWatchFor):
+                # check consecutive memory usage.
+                self_mon_info.increment_high_memory_count()
+                if (self_mon_info.corssed_memory_threshold()):
+                    # if we have crossed the memory threshold take corrective action.
+                    self.take_corrective_action(self_mon_info)
+                else:
+                    self_mon_info.reset_high_memory_count()
+            else:
+                self_mon_info.reset_high_memory_count()
+
+    def monitor_health(self):
+        """
+            Role of this function is monitor the health of the oms agent.
+            To begin with it will monitor heartbeats flowing through oms agent. 
+            We will also read oms agent logs to determine some error conditions.
+            We don't want to interfare with log watcher function.
+            So we will start this on a new thread.
+        """
+
+        self_mon_info = SelfMonitorInfo()
+        log_file_marker = LogFileMarker()
+        
+        # check every 6 minutes. we want to be bit pessimistic while looking for health, especially heartbeats which is emitted every 5 minutes.
+        sleepTime =  6 * 60  
+
+        # sleep before starting the monitoring.
+        time.sleep(sleepTime)  
+
+        while True:
+            try:
+                # Monitor heartbeat and logs.
+                self.monitor_heartbeat(self_mon_info, log_file_marker)
+
+                # Monitor memory usage
+                self.monitor_resource(self_mon_info)
+
+            except IOError as e:    
+                self._hutil_error('I/O error in monitoring health of the omsagent. Exception={0}'.format(e))
+
+            except Exception as e:
+                self._hutil_error('Error in monitoring health of the omsagent. Exception={0}'.format(e))
+
+            finally:                        
+                time.sleep(sleepTime)                     
+
+    def take_corrective_action(self, self_mon_info):
+        """
+            Take a corrective action. 
+        """
+        run_command_and_log(RestartOMSAgentServiceCommand)
```

Comment:
I don't remember seeing a lot of CRIs where simply restarting the agent fixes the issue. A better approach might be to check why the data isn't flowing. e.g. if the agent is not correctly onboarded, and try to onboard again. The ODSIngestion.status file will put out the right error message in this situation. if it's a transient error, then there isn't any need to restart.

Reply:
Infact in contrary to this i see bunch of heartbeat cri's while i was oncall which were precisely due to fluentd being in bad shape.
There are two classes of CRI's this PR handles 
1. Heartbeat stops flowing randomly due to buffer corruption/disk full [false negative] / inode [false negatuve] . Fluentd restart helps in these cases as it clears buffers or any state causing this issue.
2. Fluentd suddently stops working/ does not proceed etc. 

In both these cases control never reaches out_ods.  Fluentd's buffering layer can not forward anything to us or it just stops working. 

Obviously there is other classes of issues which we can monitor for in future but out of scope of this PR.

Reply:
Got it. I wasn't aware of this class of issues. We might also consider an update to fluentd.